### DISCUSSION TITLE
Issue with lost packets

### DISCUSSION THREAD

**Discussion by n-lebel** [libp2p/rust-libp2p#4666] [Q&A]
I’m developing a protocol that requires peers sending data to each other. To achieve that I use libp2p’s `tcp::tokio::Transport` with the `request_response` Behaviour. It’s my general understanding that the TCP transport should take care of resending lost packets: however, during unit testing, some peers sometimes get stuck indefinitely waiting for data. After implementing my own resend mechanism based on timeouts, the issue seems to be resolved, but I’m still wondering why TCP isn’t already taking care of that.

Each peer spins up a router which contains a swarm it uses to connect to others, for which the code is below. Peers run within their own `tokio::task` in unit tests, but obviously on the same machine. That definitely doesn’t fully recapitulate real-world behavior and I guess that's where the main problem lies, but I can't tell what it actually is. Any insight is welcome, thanks!

```rust
        let peer_id = PeerId::from(keypair.public());
        let transport = tcp::tokio::Transport::new(tcp::Config::default())
            .upgrade(libp2p::core::upgrade::Version::V1)
            .authenticate(NoiseConfig::new(&keypair).unwrap())
            .multiplex(MplexConfig::new())
            .boxed();

        // Create a Swarm to manage peers and events
        let request_response = request_response::cbor::Behaviour::new(
            [(
                StreamProtocol::new("/my-cbor-protocol"),
                ProtocolSupport::Full,
            )],
            request_response::Config::default(),
        );
        let mut swarm = {
            let mdns = mdns::tokio::Behaviour::new(
                mdns::Config {
                    ttl: Duration::from_secs(network_config.mdns_ttl),
                    query_interval: Duration::from_secs(network_config.mdns_query_interval),
                    enable_ipv6: network_config.mdns_ipv6,
                },
                peer_id,
            )?;
            let behaviour = CustomBehaviour {
                mdns,
                request_response,
            };
            SwarmBuilder::with_tokio_executor(transport, behaviour, peer_id).build()
        };

        // Listen on all interfaces and whatever port the OS assigns
        swarm.listen_on("/ip4/0.0.0.0/tcp/0".parse()?)?;
```

**Comment by mxinden** [comment]
Can you post your loop where you poll `swarm`? In other words, can you post your full code file?

**Reply by n-lebel** [reply]
Sure, still a WIP but here it is:

```rust
use super::{InterfaceCommand, InterfaceData, NetworkConfig, ProtocolID, RouterData};
use anyhow::Error;
use futures::{lock::Mutex, StreamExt};
use libp2p::{
    identity::Keypair,
    mdns,
    request_response::{self, ProtocolSupport},
    swarm::{NetworkBehaviour, SwarmBuilder, SwarmEvent},
    tcp, PeerId, StreamProtocol, Transport,
};
use libp2p_mplex::MplexConfig;
use libp2p_noise::Config as NoiseConfig;
use std::{collections::HashMap, sync::Arc, time::Duration};
use tokio::{
    select,
    sync::mpsc::{unbounded_channel, UnboundedSender},
};

pub type PeerNumber = usize;

#[derive(NetworkBehaviour)]
struct CustomBehaviour {
    request_response: request_response::cbor::Behaviour<RouterData, RouterData>,
    mdns: mdns::tokio::Behaviour,
}

pub struct NetworkRouter {
    peer_id: PeerId,
    network_config: NetworkConfig,
    // Needs to be able to route both requests and responses
    protocols: Mutex<HashMap<ProtocolID, UnboundedSender<InterfaceData>>>,
    // Each protocol gets a sender to send data to the router
    data_sender: UnboundedSender<InterfaceCommand>,
}

impl NetworkRouter {
    pub async fn init(network_config: NetworkConfig, keypair: Keypair) -> Result<Arc<Self>, Error> {
        // Create a random PeerId
        let peer_id = PeerId::from(keypair.public());
        info!("Local peer id: {peer_id}");
        let transport = tcp::tokio::Transport::new(tcp::Config::default())
            .upgrade(libp2p::core::upgrade::Version::V1)
            .authenticate(NoiseConfig::new(&keypair).unwrap())
            .multiplex(MplexConfig::new())
            .boxed();

        // Create a Swarm to manage peers and events
        let request_response = request_response::cbor::Behaviour::new(
            [(
                StreamProtocol::new("/my-cbor-protocol"),
                ProtocolSupport::Full,
            )],
            request_response::Config::default(),
        );
        let mut swarm = {
            let mdns = mdns::tokio::Behaviour::new(
                mdns::Config {
                    // TTL to use for mdns records. (1 day)
                    ttl: Duration::from_secs(network_config.mdns_ttl),
                    // Interval at which to poll the network for new peers.
                    // This isn’t necessary during normal operation but avoids the case that an initial packet
                    // was lost and not discovering any peers until a new peer joins the network.
                    // Receiving an mdns packet resets the timer preventing unnecessary traffic.
                    query_interval: Duration::from_secs(network_config.mdns_query_interval),
                    enable_ipv6: network_config.mdns_ipv6,
                },
                peer_id,
            )?;
            let behaviour = CustomBehaviour {
                mdns,
                request_response,
            };
            SwarmBuilder::with_tokio_executor(transport, behaviour, peer_id).build()
        };

        // Listen on all interfaces and whatever port the OS assigns
        swarm.listen_on("/ip4/0.0.0.0/tcp/0".parse()?)?;
        let protocols = Mutex::new(HashMap::<ProtocolID, UnboundedSender<InterfaceData>>::new());
        let (data_sender, mut data_receiver) = unbounded_channel::<InterfaceCommand>();
        let network_router = Arc::new(Self {
            protocols,
            peer_id,
            data_sender,
            network_config,
        });

        let network_router_clone = Arc::downgrade(&network_router);
        tokio::task::spawn(async move {
            loop {
                select! {
                        interface_command = data_receiver.recv() => {
                            if let Some(network_router_clone) = network_router_clone.upgrade() {
                                match interface_command {
                                    Some(InterfaceCommand::Send(interface_data)) => {
                                        info!("{peer_id} SENT to {}", interface_data.get_peer_id());
                                        swarm.behaviour_mut().request_response.send_request(&interface_data.get_peer_id(), RouterData::from_interface_data(interface_data));
                                    },
                                    Some(InterfaceCommand::Close(protocol_id)) => {
                                        network_router_clone.protocols.lock().await.remove(&protocol_id).expect("Attempting to delete inexistent protocol");
                                        info!("Protocol {protocol_id} finished.");
                                    },
                                    None => { return; }
                                }
                            } else {
                                println!("Router for peer id {peer_id} dropped.");
                                return;
                            }
                        }
                        event = swarm.select_next_some() => {
                            if let Some(network_router_clone) = network_router_clone.upgrade() {
                                match event  {
                                    SwarmEvent::Behaviour(CustomBehaviourEvent::RequestResponse(
                                        request_response::Event::Message { message, peer, .. },
                                    )) => match message {
                                        request_response::Message::Request {
                                            request_id: _,
                                            request,
                                            ..
                                        } => {
                                            let protocol_sender = network_router_clone.protocols.lock().await.get(&request.get_protocol_id()).unwrap().clone();
                                            protocol_sender.send(InterfaceData::from_router_data(request, peer)).unwrap();
                                            info!("RECV Request from {peer}");
                                        }
                                        request_response::Message::Response {
                                            response, request_id: _
                                        } => {
                                            let protocol_sender = network_router_clone.protocols.lock().await.get(&response.get_protocol_id()).unwrap().clone();
                                            protocol_sender.send(InterfaceData::from_router_data(response, peer)).unwrap();
                                            info!("RECV Response from {peer}");
                                        }
                                    },
                                    SwarmEvent::Behaviour(CustomBehaviourEvent::Mdns(mdns::Event::Discovered(
                                        list,
                                    ))) => {
                                        for (peer_id, _multiaddr) in list {
                                            info!("mDNS discovered a new peer: {peer_id}");
                                        }
                                    }
                                    SwarmEvent::Behaviour(CustomBehaviourEvent::Mdns(mdns::Event::Expired(
                                        list,
                                    ))) => {
                                        for (peer_id, _multiaddr) in list {
                                            info!("mDNS discovered peer has expired: {peer_id}");
                                        }
                                    }
                                    _ => {}
                                }
                            } else {
                                println!("Router for peer id {peer_id} dropped.");
                                return;
                            }
                        }
                }
            }
        });
        Ok(network_router)
    }

    pub async fn add_protocol(
        &self,
        protocol: ProtocolID,
        sender: UnboundedSender<InterfaceData>,
    ) -> Result<(), Error> {
        if self
            .protocols
            .lock()
            .await
            .insert(protocol, sender)
            .is_some()
        {
            return Err(Error::msg("Protocol ID already exists"));
        }
        Ok(())
    }
}
```

The protocols take care of the manual resending logic, although I'm planning to move that into the router directly.

**Reply by thomaseizinger** [reply]
It looks like you are implementing _another_[^1] layer of multiplexing on top of the request-response protocol. Why do you need that? You can already configure an (almost) arbitrary protocol string that should be used when you instantiate the `request_response::Behaviour`. Where is the need for dynamic protocols coming from?

In regards to your problem: TCP is "reliable" in that it will re-transmit data on an existing connection if it is not acknowledged by the remote. I don't see any connections being established in your code. You never call `Swarm::dial` to actually connect to another peer.

Some other pieces of advice:

- Please don't use mplex, it is considered deprecated. Use yamux if you have to use TCP.
- Prefer QUIC over TCP if you can, you'll get much better performance.
- Avoid adding locks to the event-loop, prefer owned data within the event-loop and use channels to transfer data in and out.
- If you are implementing your own protocols, consider creating your own `NetworkBehaviour`.

[^1]: I am saying another because libp2p already does multiplexing for you by running many protocols over the same physical connection.

**Reply by n-lebel** [reply]
Thanks a lot for your answer. I'll definitely be applying a lot of these suggestions. To answer your question, using the other layer of multiplexing stemmed from a limited understanding of the protocol and was mostly used to satisfy the type system asking for a `Boxed<PeerId, StreamMuxerBox>` type to build the swarm. Will definitely be switching to QUIC.

About the `dial` question, my understanding was that mDNS typically took care of connecting to peers as all are local in tests. The plan was to switch to explicit address dialing a bit later. Does that somehow impact the resending guarantees of TCP when communicating?

**Reply by thomaseizinger** [reply]
> To answer your question, using the other layer of multiplexing stemmed from a limited understanding of the protocol and was mostly used to satisfy the type system asking for a `Boxed<PeerId, StreamMuxerBox>` type to build the swarm.

That is not quite what I meant. Your code has a concept of `ProtocolId`s and you are using a single `request_response::Behaviour` with the protocol ID `/my-cbor-protocol` to exchange messages that seem to be namespaced by your own protocol. That is a form of multiplexing: You use a single "connection" and concurrently exchange messages for independent protocols.

I'd like to understand why you are doing this and not just create multiple instances of `request_response::Behaviour` with different protocol IDs, one for each of your usecases of exchanging messages. Can you share more about your usecase?

> About the `dial` question, my understanding was that mDNS typically took care of connecting to peers as all are local in tests. The plan was to switch to explicit address dialing a bit later. Does that somehow impact the resending guarantees of TCP when communicating?

mDNS discovers peers for you on the local network, but it doesn't make any connections by itself. That is something you need to do explicitly.

It doesn't have anything to do with TCP by itself but my theory is that you are trying to send messages to peers that you are not (yet) connected to which is why I asked about how you are connected to the other peers. Also, in the context of tests, you need to make sure peers don't get dropped to early and can actually finish sending all of their data.

Can you share you _entire_ code so we can reproduce the problem ourselves?

**Reply by n-lebel** [reply]
> I'd like to understand why you are doing this and not just create multiple instances of request_response::Behaviour with different protocol IDs, one for each of your usecases of exchanging messages. Can you share more about your usecase?

The usecase is essentially a network of nodes running multiparty computations on demand. When a computation starts, a new interface module connects to the `NetworkRouter` (partly through the channel in `protocols`) and uses it to communicate with the peers. There are several interface types, so the idea was to separate the network logic which never changes, from the protocol-specific logic which does. In that context, protocols are instantiated and dropped dynamically, hence this construct. 

> It doesn't have anything to do with TCP by itself but my theory is that you are trying to send messages to peers that you are not (yet) connected to which is why I asked about how you are connected to the other peers. Also, in the context of tests, you need to make sure peers don't get dropped to early and can actually finish sending all of their data.

Dialing the peers discovered by mDNS seems to have resolved the issue. It's interesting because the routers were still able to communicate with each other using their peer ID without dialing explicitly. The issue was sporadic, perhaps on the order of once per 10k communications, all other data being exchanged without problem. The peer just didn't receive a response to their message and simply sent it back, where it then almost always arrived after the first retry. 

> Can you share you entire code so we can reproduce the problem ourselves?

Sure, the repo is currently private but let me see if I can add you to it.

**Reply by thomaseizinger** [reply]
> > I'd like to understand why you are doing this and not just create multiple instances of request_response::Behaviour with different protocol IDs, one for each of your usecases of exchanging messages. Can you share more about your usecase?
> 
> The usecase is essentially a network of nodes running multiparty computations on demand. When a computation starts, a new interface module connects to the `NetworkRouter` (partly through the channel in `protocols`) and uses it to communicate with the peers. There are several interface types, so the idea was to separate the network logic which never changes, from the protocol-specific logic which does. In that context, protocols are instantiated and dropped dynamically, hence this construct.

Interesting, thanks for sharing!

Does this happen entirely at runtime? Would it perhaps be an option to parameterize the protocol with the two messages types instead? When you say "the protocol-specific logic changes", do you mean source-code changes? i.e. would it be fair to say that you are building yourself a networking library here that is used in multiple projects, but per project the messages stay the same?

We are currently exploring new ways of modelling request-response protocols to make them easier to use, which is why I am digging deeper here. Curious to learn how people use it and what their demands are.

> > It doesn't have anything to do with TCP by itself but my theory is that you are trying to send messages to peers that you are not (yet) connected to which is why I asked about how you are connected to the other peers. Also, in the context of tests, you need to make sure peers don't get dropped to early and can actually finish sending all of their data.
> 
> Dialing the peers discovered by mDNS seems to have resolved the issue. It's interesting because the routers were still able to communicate with each other using their peer ID without dialing explicitly. The issue was sporadic, perhaps on the order of once per 10k communications, all other data being exchanged without problem. The peer just didn't receive a response to their message and simply sent it back, where it then almost always arrived after the first retry.

This really surprises me. Where did the connections come from before you explicitly dialed the peers? How large is your network? Given that you are relying on mDNS and dialing all peers, does it mean it is small enough to run on full connectivity?

Have you considered using pub/sub instead of request-response? I don't know the nature of your multiparty computation but it could be a viable option to restructure it such that each peer emits events instead. For example, one event to share the initial parameters for the computation and then each peer emits when they are done. The benefit of this is that messages in gossipsub are relayed by other peers so even if you don't have full connectivity, each peer will eventually get each message (assuming there is _some_ route). Plus, gossipsub has topics so that could perhaps better support your needs for dynamic protocols / namespaces.

**Reply by n-lebel** [reply]
> Does this happen entirely at runtime? Would it perhaps be an option to parameterize the protocol with the two messages types instead? When you say "the protocol-specific logic changes", do you mean source-code changes? i.e. would it be fair to say that you are building yourself a networking library here that is used in multiple projects, but per project the messages stay the same?

Yes it indeed happens at runtime. Essentially a node spins up a single router, and then dynamically adds and removes different types of protocols (which are all conducted between the same peers, hence it making sense to maintain these connections alive in a single router). The routers exchange byte data, and the protocols (de)serialize this data based on their needs. The types we use don't actually implement `serde` traits, so I don't know how well they'd mesh with `request_response`. 

However if you know of a way to do dynamic multiplexing more idiomatically with libp2p I'd be very interested. I've looked a bit into dynamically modifying the protocol list within the `request_response` Behaviour, but haven't found a way so far.

> This really surprises me. Where did the connections come from before you explicitly dialed the peers? How large is your network? Given that you are relying on mDNS and dialing all peers, does it mean it is small enough to run on full connectivity?

Your guess is as good as mine. I've added you to the repo if you'd like to take a look. We don't expect the total peers to exceed 20 in production, but our testing is always done between 3 peers. I'm actually looking into just dialing the peers directly, given the list of addresses should be known at instantiation.

> Have you considered using pub/sub instead of request-response?

I will definitely look into it, but the nature of the protocols we use require some messages to be only sent between specific peers and not known to the broader network. At any given point each peer knows what they have to send and to whom. There are parts where peers broadcast data to the entire network, but since the other message types require private sending I don't know if pub/sub would yield significant improvements.

**Reply by thomaseizinger** [reply]
> > Does this happen entirely at runtime? Would it perhaps be an option to parameterize the protocol with the two messages types instead? When you say "the protocol-specific logic changes", do you mean source-code changes? i.e. would it be fair to say that you are building yourself a networking library here that is used in multiple projects, but per project the messages stay the same?
> 
> Yes it indeed happens at runtime. Essentially a node spins up a single router, and then dynamically adds and removes different types of protocols (which are all conducted between the same peers, hence it making sense to maintain these connections alive in a single router).

What is a router in your domain? Does it contain a `Swarm`? If you drop a `Swarm`, the connections will be dropped too.

> However if you know of a way to do dynamic multiplexing more idiomatically with libp2p I'd be very interested. I've looked a bit into dynamically modifying the protocol list within the request_response Behaviour, but haven't found a way so far.

Well, entirely dynamic multiplexing is a bit difficult and also not entirely meaningful. For two parties to successfully communicate with each other, they need to agree on what protocol they should speak.

If a node boots up with no knowledge of which protocols it will execute, how is the remote node going to know about these protocols? There must be some common ground here (or out-of-band communication).

> > Have you considered using pub/sub instead of request-response?
> 
> I will definitely look into it, but the nature of the protocols we use require some messages to be only sent between specific peers and not known to the broader network. At any given point each peer knows what they have to send and to whom. There are parts where peers broadcast data to the entire network, but since the other message types require private sending I don't know if pub/sub would yield significant improvements.

It actually entirely depends on how you design your systems. For example, event-based architectures don't have any point-to-point communication even though you _could_ model them as request-response. For example, from a domain perspective, you could say: "When a user orders a product, the shopping cart service needs to tell the warehouse service to ....". But the same problem can also be modeled via events where the individual producers don't know who is going to process their events.

You can apply the same thinking here: Even though individual peers know (from a domain perspective) who they need to send data to, you can also just broadcast it as an event and decouple them from each other. It might work better for your usecase, not sure :)

**Reply by n-lebel** [reply]
> What is a router in your domain? Does it contain a Swarm? If you drop a Swarm, the connections will be dropped too.

Yes it contains the Swarm, that's why it's not supposed to be dropped. Its function is to route messages to their respective protocols through the ad-hoc ProtocolId system we use. 

> Well, entirely dynamic multiplexing is a bit difficult and also not entirely meaningful. For two parties to successfully communicate with each other, they need to agree on what protocol they should speak.

There's an assumption that nodes can reach consensus on starting protocols outside of this code. They're able to agree upon a ProtocolId and a behaviour (which is always request_response). So even though they don't know that when the Swarm is created, they can all agree dynamically on adding a protocol. My question was mostly whether it's possible to move this dynamic ad-hoc ProtocolId logic directly into the Swarm (perhaps by mutating the `protocols` list in `request_response`). 

> You can apply the same thinking here: Even though individual peers know (from a domain perspective) who they need to send data to, you can also just broadcast it as an event and decouple them from each other. It might work better for your usecase, not sure :)

Interesting, thanks for clarifying, I'll definitely play around with it. I think it would be a great solution to keep in mind if it turns out we have too many peers for full connectivity.

**Reply by thomaseizinger** [reply]
> > Well, entirely dynamic multiplexing is a bit difficult and also not entirely meaningful. For two parties to successfully communicate with each other, they need to agree on what protocol they should speak.
> 
> There's an assumption that nodes can reach consensus on starting protocols outside of this code. They're able to agree upon a ProtocolId and a behaviour (which is always request_response). So even though they don't know that when the Swarm is created, they can all agree dynamically on adding a protocol. My question was mostly whether it's possible to move this dynamic ad-hoc ProtocolId logic directly into the Swarm (perhaps by mutating the `protocols` list in `request_response`).

Technically yes, it could be expressed as a behaviour. I've actually built stuff like this in the past but it turns out to be somewhat broken. The problem is that dynamically registering protocol handlers creates somewhat weird timing constraints for a system. You need to ensure that a node has its protocol handler registered before another one starts talking to it for example.

Typically, you are much better served by designing your system such that you can register your protocol / message handlers once on startup and then just deal with incoming messages as they arrive.

Can you give me a more concrete example of what you are trying to build? This all still sounds very abstract to me which makes it difficult to think of a different way of modelling this.

**Reply by n-lebel** [reply]
That makes sense. The idea was that since nodes are able to agree beforehand on starting a protocol (and therefore wait until everyone has registered their respective handlers), these types of issues shouldn't arise. I tended to favor this approach because I have a pool of nodes from which specific subsets are chosen to run multiparty computations on demand. It seemed more appropriate to have a structure that just handles the connections and the raw sending/receiving (which I refer to as a `Router`, since it routes incoming messages to the appropriate MPC protocol), because it made little sense to me to have to reconnect to all the peers every time you wanted to run a protocol. 

The other reason for having separate structs is that there are different kinds of protocols that all require network communication (provided by one single `Router`) as well as protocol-specific higher-level functions (provided by `NetworkInterface` structs that are specific to each protocol type). Also more practically, there were generic types in my construct which weren't `Serialize + Deserialize` (but rather arkworks' own `CanonicalSerialize + CanonicalDeserialize` traits), which kept running into issues with libp2p. 

So the logical conclusion for me was to decouple the generic network functionalities (connecting to peers and sending/receiving bytes, essentially) which can be instantiated once on startup, from MPC protocols which need to be instantiated dynamically. Hope that makes sense, happy to walk you through the code more in detail if it doesn't. I've included a short diagram, might make it easier to visualize. 

![IMG_96332B579756-1](https://github.com/libp2p/rust-libp2p/assets/126451245/dc70d2bd-7169-4225-8071-41e5bbc6784f)

**Reply by thomaseizinger** [reply]
The recently released `stream` protocol might be of interest to you: https://github.com/libp2p/rust-libp2p/tree/master/protocols/stream
